{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NC_007795.1:c74979-73429\n",
      "Name: NC_007795.1:c74979-73429\n",
      "Description: NC_007795.1:c74979-73429 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome\n",
      "Number of features: 0\n",
      "Seq('TTGAAAAAGAAAAACATTTATTCAATTCGTAAACTAGGTGTAGGTATTGCATCT...TAA') \n",
      "\n",
      "\n",
      "Sequence: TTGAAAAAGAAAAACATTTATTCAATTCGTAAACTAGGTGTAGGTATTGCATCTGTAACTTTAGGTACATTACTTATATCTGGTGGCGTAACACCTGCTGCAAATGCTGCGCAACACGATGAAGCTCAACAAAATGCTTTTTATCAAGTCTTAAATATGCCTAACTTAAATGCTGATCAACGCAATGGTTTTATCCAAAGCCTTAAAGATGATCCAAGCCAAAGTGCTAACGTTTTAGGTGAAGCTCAAAAACTTAATGACTCTCAAGCTCCAAAAGCTGATGCGCAACAAAATAACTTCAACAAAGATCAACAAAGCGCCTTCTATGAAATCTTGAACATGCCTAACTTAAACGAAGCGCAACGTAACGGCTTCATTCAAAGTCTTAAAGACGACCCAAGCCAAAGCACTAACGTTTTAGGTGAAGCTAAAAAATTAAACGAATCTCAAGCACCGAAAGCTGATAACAATTTCAACAAAGAACAACAAAATGCTTTCTATGAAATCTTGAATATGCCTAACTTAAACGAAGAACAACGCAATGGTTTCATCCAAAGCTTAAAAGATGACCCAAGCCAAAGTGCTAACCTATTGTCAGAAGCTAAAAAGTTAAATGAATCTCAAGCACCGAAAGCGGATAACAAATTCAACAAAGAACAACAAAATGCTTTCTATGAAATCTTACATTTACCTAACTTAAACGAAGAACAACGCAATGGTTTCATCCAAAGCCTAAAAGATGACCCAAGCCAAAGCGCTAACCTTTTAGCAGAAGCTAAAAAGCTAAATGATGCTCAAGCACCAAAAGCTGACAACAAATTCAACAAAGAACAACAAAATGCTTTCTATGAAATTTTACATTTACCTAACTTAACTGAAGAACAACGTAACGGCTTCATCCAAAGCCTTAAAGACGATCCTTCAGTGAGCAAAGAAATTTTAGCAGAAGCTAAAAAGCTAAACGATGCTCAAGCACCAAAAGAGGAAGACAATAACAAGCCTGGCAAAGAAGACAATAACAAGCCTGGCAAAGAAGACAACAACAAGCCTGGTAAAGAAGACAACAACAAGCCTGGTAAAGAAGACAACAACAAGCCTGGCAAAGAAGACGGCAACAAGCCTGGTAAAGAAGACAACAAAAAACCTGGTAAAGAAGATGGCAACAAGCCTGGTAAAGAAGACAACAAAAAACCTGGTAAAGAAGACGGCAACAAGCCTGGCAAAGAAGATGGCAACAAACCTGGTAAAGAAGATGGTAACGGAGTACATGTCGTTAAACCTGGTGATACAGTAAATGACATTGCAAAAGCAAACGGCACTACTGCTGACAAAATTGCTGCAGATAACAAATTAGCTGATAAAAACATGATCAAACCTGGTCAAGAACTTGTTGTTGATAAGAAGCAACCAGCAAACCATGCAGATGCTAACAAAGCTCAAGCATTACCAGAAACTGGTGAAGAAAATCCATTCATCGGTACAACTGTATTTGGTGGATTATCATTAGCCTTAGGTGCAGCGTTATTAGCTGGACGTCGTCGCGAACTATAA\n",
      "\n",
      "Complement: AACTTTTTCTTTTTGTAAATAAGTTAAGCATTTGATCCACATCCATAACGTAGACATTGAAATCCATGTAATGAATATAGACCACCGCATTGTGGACGACGTTTACGACGCGTTGTGCTACTTCGAGTTGTTTTACGAAAAATAGTTCAGAATTTATACGGATTGAATTTACGACTAGTTGCGTTACCAAAATAGGTTTCGGAATTTCTACTAGGTTCGGTTTCACGATTGCAAAATCCACTTCGAGTTTTTGAATTACTGAGAGTTCGAGGTTTTCGACTACGCGTTGTTTTATTGAAGTTGTTTCTAGTTGTTTCGCGGAAGATACTTTAGAACTTGTACGGATTGAATTTGCTTCGCGTTGCATTGCCGAAGTAAGTTTCAGAATTTCTGCTGGGTTCGGTTTCGTGATTGCAAAATCCACTTCGATTTTTTAATTTGCTTAGAGTTCGTGGCTTTCGACTATTGTTAAAGTTGTTTCTTGTTGTTTTACGAAAGATACTTTAGAACTTATACGGATTGAATTTGCTTCTTGTTGCGTTACCAAAGTAGGTTTCGAATTTTCTACTGGGTTCGGTTTCACGATTGGATAACAGTCTTCGATTTTTCAATTTACTTAGAGTTCGTGGCTTTCGCCTATTGTTTAAGTTGTTTCTTGTTGTTTTACGAAAGATACTTTAGAATGTAAATGGATTGAATTTGCTTCTTGTTGCGTTACCAAAGTAGGTTTCGGATTTTCTACTGGGTTCGGTTTCGCGATTGGAAAATCGTCTTCGATTTTTCGATTTACTACGAGTTCGTGGTTTTCGACTGTTGTTTAAGTTGTTTCTTGTTGTTTTACGAAAGATACTTTAAAATGTAAATGGATTGAATTGACTTCTTGTTGCATTGCCGAAGTAGGTTTCGGAATTTCTGCTAGGAAGTCACTCGTTTCTTTAAAATCGTCTTCGATTTTTCGATTTGCTACGAGTTCGTGGTTTTCTCCTTCTGTTATTGTTCGGACCGTTTCTTCTGTTATTGTTCGGACCGTTTCTTCTGTTGTTGTTCGGACCATTTCTTCTGTTGTTGTTCGGACCATTTCTTCTGTTGTTGTTCGGACCGTTTCTTCTGCCGTTGTTCGGACCATTTCTTCTGTTGTTTTTTGGACCATTTCTTCTACCGTTGTTCGGACCATTTCTTCTGTTGTTTTTTGGACCATTTCTTCTGCCGTTGTTCGGACCGTTTCTTCTACCGTTGTTTGGACCATTTCTTCTACCATTGCCTCATGTACAGCAATTTGGACCACTATGTCATTTACTGTAACGTTTTCGTTTGCCGTGATGACGACTGTTTTAACGACGTCTATTGTTTAATCGACTATTTTTGTACTAGTTTGGACCAGTTCTTGAACAACAACTATTCTTCGTTGGTCGTTTGGTACGTCTACGATTGTTTCGAGTTCGTAATGGTCTTTGACCACTTCTTTTAGGTAAGTAGCCATGTTGACATAAACCACCTAATAGTAATCGGAATCCACGTCGCAATAATCGACCTGCAGCAGCGCTTGATATT\n",
      "\n",
      "Reverse Complement: TTATAGTTCGCGACGACGTCCAGCTAATAACGCTGCACCTAAGGCTAATGATAATCCACCAAATACAGTTGTACCGATGAATGGATTTTCTTCACCAGTTTCTGGTAATGCTTGAGCTTTGTTAGCATCTGCATGGTTTGCTGGTTGCTTCTTATCAACAACAAGTTCTTGACCAGGTTTGATCATGTTTTTATCAGCTAATTTGTTATCTGCAGCAATTTTGTCAGCAGTAGTGCCGTTTGCTTTTGCAATGTCATTTACTGTATCACCAGGTTTAACGACATGTACTCCGTTACCATCTTCTTTACCAGGTTTGTTGCCATCTTCTTTGCCAGGCTTGTTGCCGTCTTCTTTACCAGGTTTTTTGTTGTCTTCTTTACCAGGCTTGTTGCCATCTTCTTTACCAGGTTTTTTGTTGTCTTCTTTACCAGGCTTGTTGCCGTCTTCTTTGCCAGGCTTGTTGTTGTCTTCTTTACCAGGCTTGTTGTTGTCTTCTTTACCAGGCTTGTTGTTGTCTTCTTTGCCAGGCTTGTTATTGTCTTCTTTGCCAGGCTTGTTATTGTCTTCCTCTTTTGGTGCTTGAGCATCGTTTAGCTTTTTAGCTTCTGCTAAAATTTCTTTGCTCACTGAAGGATCGTCTTTAAGGCTTTGGATGAAGCCGTTACGTTGTTCTTCAGTTAAGTTAGGTAAATGTAAAATTTCATAGAAAGCATTTTGTTGTTCTTTGTTGAATTTGTTGTCAGCTTTTGGTGCTTGAGCATCATTTAGCTTTTTAGCTTCTGCTAAAAGGTTAGCGCTTTGGCTTGGGTCATCTTTTAGGCTTTGGATGAAACCATTGCGTTGTTCTTCGTTTAAGTTAGGTAAATGTAAGATTTCATAGAAAGCATTTTGTTGTTCTTTGTTGAATTTGTTATCCGCTTTCGGTGCTTGAGATTCATTTAACTTTTTAGCTTCTGACAATAGGTTAGCACTTTGGCTTGGGTCATCTTTTAAGCTTTGGATGAAACCATTGCGTTGTTCTTCGTTTAAGTTAGGCATATTCAAGATTTCATAGAAAGCATTTTGTTGTTCTTTGTTGAAATTGTTATCAGCTTTCGGTGCTTGAGATTCGTTTAATTTTTTAGCTTCACCTAAAACGTTAGTGCTTTGGCTTGGGTCGTCTTTAAGACTTTGAATGAAGCCGTTACGTTGCGCTTCGTTTAAGTTAGGCATGTTCAAGATTTCATAGAAGGCGCTTTGTTGATCTTTGTTGAAGTTATTTTGTTGCGCATCAGCTTTTGGAGCTTGAGAGTCATTAAGTTTTTGAGCTTCACCTAAAACGTTAGCACTTTGGCTTGGATCATCTTTAAGGCTTTGGATAAAACCATTGCGTTGATCAGCATTTAAGTTAGGCATATTTAAGACTTGATAAAAAGCATTTTGTTGAGCTTCATCGTGTTGCGCAGCATTTGCAGCAGGTGTTACGCCACCAGATATAAGTAATGTACCTAAAGTTACAGATGCAATACCTACACCTAGTTTACGAATTGAATAAATGTTTTTCTTTTTCAA\n",
      "\n",
      "Length: 1551 bp\n",
      "\n",
      "GC Content: 38.168923275306256%\n",
      "\n",
      "Nucleotide Frequency:\n",
      "A: 0.41070277240490005\n",
      "T: 0.2076079948420374\n",
      "C: 0.20438426821405545\n",
      "G: 0.1773049645390071\n",
      "\n",
      "Dinucleotide Frequency:\n",
      "AA: 0.20451612903225808\n",
      "AT: 0.05870967741935484\n",
      "AC: 0.07548387096774194\n",
      "AG: 0.07161290322580645\n",
      "TA: 0.06\n",
      "TT: 0.05741935483870968\n",
      "TC: 0.03483870967741935\n",
      "TG: 0.05548387096774193\n",
      "CA: 0.08709677419354839\n",
      "CT: 0.06\n",
      "CC: 0.032903225806451615\n",
      "CG: 0.024516129032258065\n",
      "GA: 0.05935483870967742\n",
      "GT: 0.03096774193548387\n",
      "GC: 0.06129032258064516\n",
      "GG: 0.025806451612903226\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "file = \"S-aureus-NCTC8325.fasta\"\n",
    "\n",
    "record_iterator = SeqIO.parse(file, \"fasta\")\n",
    "first_record = next(record_iterator)\n",
    "print(first_record,\"\\n\")\n",
    "\n",
    "for seq_rec in SeqIO.parse(file, \"fasta\"):\n",
    "    seq = seq_rec.seq\n",
    "    l = len(seq)\n",
    "    print(\"\\nSequence:\",seq)\n",
    "    print(\"\\nComplement:\",seq.complement())\n",
    "    print(\"\\nReverse Complement:\",seq.reverse_complement())\n",
    "    print(f\"\\nLength: {l} bp\")\n",
    "    print(f\"\\nGC Content: {GC(seq)}%\")\n",
    "\n",
    "A = seq.count(\"A\")\n",
    "T = seq.count(\"T\")\n",
    "C = seq.count(\"C\")\n",
    "G = seq.count(\"G\")\n",
    "Ap = A/l\n",
    "Tp = T/l\n",
    "Cp = C/l\n",
    "Gp = G/l\n",
    "print(f\"\\nNucleotide Frequency:\\nA: {Ap}\\nT: {Tp}\\nC: {Cp}\\nG: {Gp}\")\n",
    "\n",
    "AA = 0\n",
    "AT = 0\n",
    "AC = 0\n",
    "AG = 0\n",
    "TA = 0\n",
    "TT = 0\n",
    "TC = 0\n",
    "TG = 0\n",
    "CA = 0\n",
    "CT = 0\n",
    "CC = 0\n",
    "CG = 0\n",
    "GA = 0\n",
    "GT = 0\n",
    "GC = 0\n",
    "GG = 0\n",
    "bp_num = 0\n",
    "\n",
    "for bp in seq:\n",
    "    if seq[bp_num]+seq[bp_num+1] == \"AA\":\n",
    "        AA += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"AT\":\n",
    "        AT += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"AC\":\n",
    "        AC+= 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"AG\":\n",
    "        AG += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"TA\":\n",
    "        TA += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"TT\":\n",
    "        TT += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"TC\":\n",
    "        TC+= 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"TG\":\n",
    "        TG += 1\n",
    "    elif seq[bp_num]+seq[bp_num + 1] == \"CA\":\n",
    "        CA += 1\n",
    "    elif seq[bp_num]+seq[bp_num + 1] == \"CT\":\n",
    "        CT += 1\n",
    "    elif seq[bp_num]+seq[bp_num + 1] == \"CC\":\n",
    "        CC += 1\n",
    "    elif seq[bp_num]+seq[bp_num + 1] == \"CG\":\n",
    "        CG += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"GA\":\n",
    "        GA += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"GT\":\n",
    "        GT += 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"GC\":\n",
    "        GC+= 1\n",
    "    elif seq[bp_num]+seq[bp_num+1] == \"GG\":\n",
    "        GG += 1\n",
    "    bp_num += 1\n",
    "    if bp_num == l - 1:\n",
    "        break\n",
    "\n",
    "dinuc_count = AA+AT+AC+AG+TA+TT+TC+TG+CA+CT+CC+CG+GA+GT+GC+GG\n",
    "AAp = AA/dinuc_count\n",
    "ATp = AT/dinuc_count\n",
    "ACp = AC/dinuc_count\n",
    "AGp = AG/dinuc_count\n",
    "TAp = TA/dinuc_count\n",
    "TTp = TT/dinuc_count\n",
    "TCp = TC/dinuc_count\n",
    "TGp = TG/dinuc_count\n",
    "CAp = CA/dinuc_count\n",
    "CTp = CT/dinuc_count\n",
    "CCp = CC/dinuc_count\n",
    "CGp = CG/dinuc_count\n",
    "GAp = GA/dinuc_count\n",
    "GTp = GT/dinuc_count\n",
    "GCp = GC/dinuc_count\n",
    "GGp = GG/dinuc_count\n",
    "print(f\"\\nDinucleotide Frequency:\\nAA: {AAp}\\nAT: {ATp}\\nAC: {ACp}\\nAG: {AGp}\\nTA: {TAp}\\nTT: {TTp}\\nTC: {TCp}\\nTG: {TGp}\\nCA: {CAp}\\nCT: {CTp}\\nCC: {CCp}\\nCG: {CGp}\\nGA: {GAp}\\nGT: {GTp}\\nGC: {GCp}\\nGG: {GGp}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
